### PR TITLE
Begin process of removing singleton use by StringMatcher

### PR DESCRIPTION
**Pull request by ggreenway**
Commit Message: Begin process of removing singleton use by StringMatcher

This PR temporarily adds two versions of StringMatcherImpl: one that uses the existing singletons, and one that takes a factory context. 

Subsequent PRs will gradually convert all uses of StringMatcherImpl to use the factory context. Then the version using singletons will be deleted, along with the code that sets the singletons.
Additional Description: This work is being broken into multiple PRs because otherwise it would be very large and difficult to review.

Part of #32792
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by RyanTheOptimist — COMMENTED** [review]
Looks reasonable, but I think ASAN is finding some correctness issues, eg:

source/common/common/matchers.h:104:66: runtime error: reference binding to null pointer of type 'Regex::Engine'

**Comment by ggreenway** [comment]
> source/common/common/matchers.h:104:66: runtime error: reference binding to null pointer of type 'Regex::Engine'

Yep, some instances where I converted a nullptr to a reference. The reference was never used, but still gets flagged by asan. I think I've worked around it now, but waiting on CI. 

There was also a real failure on mobile that I think I've fixed.

**Review by RyanTheOptimist — COMMENTED** [review]
LGTM, modulo coverage. But maybe that's hard to fix?

File: test/per_file_coverage.sh
Comment (ggreenway) [reply test/per_file_coverage.sh:66]: I added `regexEngine()` to the `Server::Instance`. This isn't used (yet) in the config_validation path, because most of the uses of `StringMatcherImpl` still don't use it. Future PRs will convert all of those, and this line will get hit in coverage at that time.

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
